plugins/askrene/refine.c

```diff
@@ -597,29 +598,42 @@ double flows_probability(const tal_t *ctx, struct route_query *rq,
 	return probability;
 }
 
-/* Compare flows by deliver amount */
-static int reverse_cmp_flows(struct flow *const *fa, struct flow *const *fb,
-			     void *unused UNUSED)
+/* Compare flows by remaining capacity (note: they are not reserved, so this
+ * capacity includes current capacity. */
+static int reverse_cmp_capacity(struct flow *const *fa, struct flow *const *fb,
+				struct route_query *rq)
 {
-	if (amount_msat_eq((*fa)->delivers, (*fb)->delivers))
+	struct amount_msat amount_a, amount_b;
+
+	amount_a = flow_max_deliverable(rq, *fa, NULL);
+	amount_b = flow_max_deliverable(rq, *fb, NULL);
+	if (!amount_msat_deduct(&amount_a, (*fa)->delivers))
+		abort();
+	if (!amount_msat_deduct(&amount_b, (*fb)->delivers))
+		abort();
+
+	if (amount_msat_eq(amount_a, amount_b))
 		return 0;
-	if (amount_msat_greater((*fa)->delivers, (*fb)->delivers))
+	if (amount_msat_greater(amount_a, amount_b))
 		return -1;
 	return 1;
 }
 
-bool remove_flows(struct flow ***flows, u32 n)
+const char *reduce_num_flows(const tal_t *ctx,
+			     const struct route_query *rq,
+			     struct flow ***flows,
+			     struct amount_msat deliver,
+			     size_t num_parts)
 {
-	if (n == 0)
-		goto fail;
-	if (n > tal_count(*flows))
-		goto fail;
-	asort(*flows, tal_count(*flows), reverse_cmp_flows, NULL);
-	for (size_t count = tal_count(*flows); n > 0; n--, count--) {
-		assert(count > 0);
-		tal_arr_remove(flows, count - 1);
-	}
-	return true;
-fail:
-	return false;
+	/* Keep the largest-capacity flows */
+	size_t orig_num_flows = tal_count(*flows);
+	asort(*flows, orig_num_flows, reverse_cmp_capacity, cast_const(struct route_query *, rq));
```

Comment:
I think it is better to keep the largest-deliver flows.
If we only go for the highest capacity we may throw away the low cost benefits of the MCF.
Hypothetical scenario: MCF finds 3 flows but maxparts=2,
flow 1: deliver=10, cost=0, capacity=0
flow 2: deliver=7, cost=1, capacity=5
flow 3: deliver=1, cost=10, capacity=100
It is better to keep flows 1 and 2 by accomodating 1 more unit of flow in flow2 at 1 value expense (per flow),
than to keep flows 2 and 3 by accomodating 5 more units of flow in flow2 at cost 1 and 5 in flow3 at cost 100.

Reply:
Convinced.  It's also simpler to implement :)